FW/Presentations/Lab_Guides/_L_C_04_Writing_UEFI_App_Simics_Linux_Lab.md

```diff
@@ -0,0 +1,2024 @@
+FS1:\>
+```
+
+5. Exit Simics
+
+```
+simics> stop
+simics> quit
+```
+
+---
+## Slide 63 Lab 5: Bug Reports
+
+You received the following bug reports on your SampleApp UEFI Application
+
+1. If an "Enter" Key is pressed before the "." character, there is no line feed to go to a new line like a real typewriter would do.
+2. Since the DOT "." is used in sentences a different character other than a "." is required to exit the Typewriter function. One suggestion was to use the "ESC" character instead.
+3. Some customers would like the application without the typewriter feature enabled.
+
+How would you add these fixed to the SampleApp UEFI application?
+
+---
+## Slide 64 Lab 5.1: Bug Fixes
+
+Add a line feed character after a "Enter" key.
+
+1. Update SampleApp.c with this fix (hint: print a CHAR_LINEFEED when a Carriage return is entered)
+2. At the Terminal Command Prompt, Re-Build BoardX58Ich10
+
+```bash
+$> cd ~/fw/edk2-ws/edk2-platforms/Platform/Intel/
+$> python build_bios.py -p BoardX58Ich10 -t GCC5
+```
+
+3. Copy SampleApp.efi from the build directory to the VHD Disk
+
+```bash
+$> cp ../Build/SimicsOpenBoardPkg/BoardX58Ich10/DEBUG_GCC5/SampleApp.efi ~/VHD
+```
+
+4. Run the qsp-modern-core script from Terminal Prompt:
+
+```bash
+$> ./simics targets/qsp-x86/qsp-modern-core.simics
+simics> run
+```
+
+```
+Shell> Fs1:
+FS1:/> SampleApp
+System Table: 0x0DEFED018
+
+Press any key to continue:
+Enter text. Include a dot ('.') in a sentence then <Enter> to exit:
+This is the First Line
+This is the second Line.
+
+FS1:\>
+```
+
+5. Exit Simics
+
+```
+simics> stop
+simics> quit
+```
+
+---
+## Slide 65 Lab 5.2: Bug Fixes
+
+Use the Scan Code for ESC instead of the DOT character to exit the Typewriter Function (Hint: the ExitLoop flag will no longer be needed. Hint: Search workspace for SCAN_ESC)
+
+1. Update SampleApp.c to use SCAN_ESC to exit the do-loop
+2. At the Terminal Command Prompt, Re-Build BoardX58Ich10
+
+```bash
+$> cd ~/fw/edk2-ws/edk2-platforms/Platform/Intel/
+$> python build_bios.py -p BoardX58Ich10 -t GCC5
+```
+
+3. Copy SampleApp.efi from the build directory to the VHD Disk
+
+```bash
+$> cp ../Build/SimicsOpenBoardPkg/BoardX58Ich10/DEBUG_GCC5/SampleApp.efi ~/VHD
+```
+
+4. Run the qsp-modern-core script from Terminal Prompt:
+
+```bash
+$> ./simics targets/qsp-x86/qsp-modern-core.simics
+simics> run
+```
+
+```
+Shell> Fs1:
+FS1:/> SampleApp
+System Table: 0x0DEFED018
+
+Press any key to continue:
+Enter text as in a typewriter then type <ESC> to exit
+This is the First Line
+This is the second Line.    Now press "ESC" Key
+FS1:\>
+```
+
+5. Exit Simics
+
+```
+simics> stop
+simics> quit
+```
+
+---
+## Slide 66 Lab 5.3: Bug Fixes
+
+Make the PCD PcdTypeWriterFeatureEnable determined by a Build Flag
+
+1. Update OpenBoardPkg.dsc (about line 45)
+
+```
+DEFINE ADD_TYPEWRITER       = TRUE
+```
+
+2. Update OpenBoardPkgPcd.dsc (at end in [PcdsFeatureFlag] Section)
+
```

Comment:
2.  Update OpenBoardPkgPcd.dsc  (at end add  [Packages]   MyPkg/MyPkg.dec  and  [PcdsFeatureFlag] Section)
```
[Packages]
  MyPkg/MyPkg.dec 

[PcdsFeatureFlag]
!if   $(ADD_TYPEWRITTER) == TRUE
    gMyPkgTokenSpaceGuid.PcdTypeWriterFeatureEnable|TRUE
!else
   gMyPkgTokenSpaceGuid.PcdTypeWriterFeatureEnable|FALSE
!endif
```